Changes for pycurl, Requests and service classes to use pycurl by default (do not merge)

vkuznet opened the pull request.

vkuznet committed:
Changes to make pycurl plugin usable in WMCore.

vkuznet commented:
Eric,
why the build on jenkins page has a red ball even though
DMWM-PR-unittests is yellow? From the console output I see the following:

```
Checking console output
/var/lib/jenkins/jobs/DMWM-PR-test/builds/1024/log:
Testing of python code. DMWM-FAIL-UNIT
Run condition [Text Finder] enabling perform for step [BuilderChain]
[DMWM-PR-test] $ /bin/sh -xe /tmp/hudson4038148752549809285.sh
+ echo 'One or more checks gave bad results'
One or more checks gave bad results
+ exit 1
```

I'm not sure which DMWM-FAIL-UNIT it reads.


On  0, DMWM Jenkins <notifications@github.com> wrote:
> Beginning pylint and unit tests.
> See https://cmssdt.cern.ch/jenkins/job/DMWM-PR-test/1024/ for details
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-151566483

ericvaandering commented:
You have a red ball for two reasons:

One unit test moved from success to error
Score for src/python/WMCore/Services/pycurl_manager.py changed from 9.72 to 9.68

All the yellow ball for DMWM-unittests means is that it was able to run all the tests (no timeouts or other crashes in the script). DMWM-PR-test does the final analysis of all the sub jobs and determines success or failure.

vkuznet commented:
test this please

vkuznet commented:
Sengchan, it seems to me that all tests passed, but I found that two latests builds 1033 and 1034 contradict each other in terms of unit tests. It seems to me that those are unstable tests. Could you please check. Apart from that I think pycurl is ready to go. Please review the changes and let me know if you want to merge them or split them apart. This PR can super seed #5845 because it contains its changes plus additional one I made.

ticoann commented:
Thank you very much Valentin, yes those tests are unstable test. I will close #5845 then.  Changes looks good to me, yes, please merge commits together.  I will merge this patch after we update the production agents. Thanks again

vkuznet commented:
Seangchan, I squashed the changes, please review and let's wait for unit tests to complete.

vkuznet commented:
test this please

ticoann commented:
@vkuznet Hi Valentin, I getting some errors with this patch. (*)

Error seems to be happening here.
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WebTools/RESTModel.py#L246

I will try to create the unittest which we can reproduce the problem. (Unless you know what where it causes the problem, please don't investigate yet)

(*)
2015-11-17 18:05:15,093 WARNING  Service: The cachefile /data/srv/state/workqueue/cache/.wmcore_cache_100034/requests/reqmgr2-dev.cern.ch/-1062535079034185883_PUT_request does not exist and the service at https://reqmgr2-dev.cern.ch/reqmgr/reqMgr/request is unavailable - it returned 400 because Bad Request
 with result: {"exception": 400, "message": "Invalid input: Input arguments failed sanitation. XXXXX", "type": "HTTPError"}

2015-11-17 18:05:15,220 ERROR    WorkQueueReqMgrInterface: Error talking to ReqMgr about request "sryu_MonteCarloFromGEN_pnn_test_151117_061147_4617": Traceback (most recent call last):
  File "/data/srv/HG1511c/sw.pre/slc6_amd64_gcc481/cms/workqueue/1.0.11.pre8/lib/python2.6/site-packages/WMCore/WorkQueue/WorkQueueReqMgrInterface.py", line 191, in report
    self.reportElement(ele)
  File "/data/srv/HG1511c/sw.pre/slc6_amd64_gcc481/cms/workqueue/1.0.11.pre8/lib/python2.6/site-packages/WMCore/WorkQueue/WorkQueueReqMgrInterface.py", line 321, in reportElement
    self.reportRequestStatus(element['RequestName'], element['Status'])
  File "/data/srv/HG1511c/sw.pre/slc6_amd64_gcc481/cms/workqueue/1.0.11.pre8/lib/python2.6/site-packages/WMCore/WorkQueue/WorkQueueReqMgrInterface.py", line 275, in reportRequestStatus
    raise ex
HTTPException: url=https://reqmgr2-dev.cern.ch/reqmgr/reqMgr/request, code=400, reason=Bad Request, headers={'Content-Length': '109', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Connection': 'close', 'Etag': '4604053556711648119', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Tue, 17 Nov 2015 17:05:15 GMT', 'Content-Type': 'application/json', 'CMS-Server-Time': 'D=8306 t=1447779915082738'}

vkuznet commented:
Can you post how to reproduce this problem. The raised exception does not
provide any information about which arguments have failed, so I would suggest
to fix exception message to give us more information.

On  0, ticoann <notifications@github.com> wrote:
> @vkuznet Hi Valentin, I getting some errors with this patch. (*)
> 
> Error seems to be happening here.
> https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WebTools/RESTModel.py#L246
> 
> I will try to create the unittest which we can reproduce the problem. (Unless you know what where it causes the problem, please don't investigate yet)
> 
> (*)
> 2015-11-17 18:05:15,093 WARNING  Service: The cachefile /data/srv/state/workqueue/cache/.wmcore_cache_100034/requests/reqmgr2-dev.cern.ch/-1062535079034185883_PUT_request does not exist and the service at https://reqmgr2-dev.cern.ch/reqmgr/reqMgr/request is unavailable - it returned 400 because Bad Request
>  with result: {"exception": 400, "message": "Invalid input: Input arguments failed sanitation. XXXXX", "type": "HTTPError"}
> 
> 2015-11-17 18:05:15,220 ERROR    WorkQueueReqMgrInterface: Error talking to ReqMgr about request "sryu_MonteCarloFromGEN_pnn_test_151117_061147_4617": Traceback (most recent call last):
>   File "/data/srv/HG1511c/sw.pre/slc6_amd64_gcc481/cms/workqueue/1.0.11.pre8/lib/python2.6/site-packages/WMCore/WorkQueue/WorkQueueReqMgrInterface.py", line 191, in report
>     self.reportElement(ele)
>   File "/data/srv/HG1511c/sw.pre/slc6_amd64_gcc481/cms/workqueue/1.0.11.pre8/lib/python2.6/site-packages/WMCore/WorkQueue/WorkQueueReqMgrInterface.py", line 321, in reportElement
>     self.reportRequestStatus(element['RequestName'], element['Status'])
>   File "/data/srv/HG1511c/sw.pre/slc6_amd64_gcc481/cms/workqueue/1.0.11.pre8/lib/python2.6/site-packages/WMCore/WorkQueue/WorkQueueReqMgrInterface.py", line 275, in reportRequestStatus
>     raise ex
> HTTPException: url=https://reqmgr2-dev.cern.ch/reqmgr/reqMgr/request, code=400, reason=Bad Request, headers={'Content-Length': '109', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Connection': 'close', 'Etag': '4604053556711648119', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Tue, 17 Nov 2015 17:05:15 GMT', 'Content-Type': 'application/json', 'CMS-Server-Time': 'D=8306 t=1447779915082738'}
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-157450061

ticoann commented:
Actually the code I pointed is where the exception occur. I will try to create the unittest reproduce the problem so it can be investigated. (Please don't spend time until then, thanks)
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WebTools/RESTModel.py#L246

vkuznet commented:
Or, you can fix RESTModel.py#L246 in your local setup to print kwars e.g.

change
        if input_kwargs:
            raise HTTPError(400, 'Invalid input: Input arguments failed sanitation.')

to
        if input_kwargs:
            print "Invalid kwargs", input_kwargs
            raise HTTPError(400, 'Invalid input: Input arguments failed sanitation.')

and then re-run what you had before.

On  0, ticoann <notifications@github.com> wrote:
> Actually the code I pointed is where the exception occur. I will try to create the unittest reproduce the problem so it can be investigated. (Please don't spend time until then, thanks)
> https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WebTools/RESTModel.py#L246
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-157484808

ticoann commented:
I did already, somehow that reads to 500 error some where else.

ticoann commented:
Hi Valentin, it seems the content length is needed when make post request (or request body is not empty). (Our webtool requirement) You can run below unittest 
https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py

vkuznet committed:
Define POSTFIELDS in PUT request even for empty parameters, this trigger setup of Content-Length

vkuznet commented:
Seangchan,
can you re-run your tests now. It seems tome I fixed 411 error, but I'm facing
another one

{"exception": 400, "message": "Invalid input: Input data failed validation.",
"type": "HTTPError"}

which came from 

debug(2): PUT /unittests/rest/version/fake/fake HTTP/1.1
Host: localhost:8899
Cache-Control: no-cache
Content-type: application/json
Accept: application/json
User-agent: WMCore.Services.Requests/v001
Content-Length: 0


WARNING:cherrypy.access:NullAuth called for:
WARNING:cherrypy.access:    role(s): []
    group(s): []
    site(s): []
DEBUG:cherrypy.error:version raw data: {'args': [], 'kwargs': {}}
DEBUG:cherrypy.error:version sanitised input_data: {'version': 'fake', 'scramArch': 'fake'}
DEBUG:cherrypy.error:'fake' does not match regular expression CMSSW(_\d+){3}(_[a-zA-Z0-9_]+)?$
DEBUG:cherrypy.error:call to PUT with args: ['version', 'fake', 'fake'] kwargs: {} resulted in Invalid input: Input data failed validation.
INFO:cherrypy.access:[18/Nov/2015:11:06:32] lnxcu9.classe.cornell.edu 127.0.0.1 "PUT /unittests/rest/version/fake/fake HTTP/1.1" 400 Bad Request [data: - in 98 out 1166 us ] [auth: - "" "" ] [ref: "" "WMCore.Services.Requests/v001" ]
debug(1): HTTP/1.1 400 Bad Request

as you can see this is not HTTP issue per se, and rather validation of PUT request
fake does not match regular expression CMSSW(_\d+){3}(_[a-zA-Z0-9_]+)?$

So, I'm not sure it this is something I need to fix in this PR.

On  0, ticoann <notifications@github.com> wrote:
> Hi Valentin, it seems the content length is needed when make post request (or request body is not empty). (Our webtool requirement) You can run below unittest 
> https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-157514972

ticoann commented:
Thanks Valentin, I will rerun the integration test.

ticoann commented:
test this please

ticoann commented:
test this please

ticoann commented:
test this please

ticoann commented on src/python/WMCore/Services/Requests.py:
```diff
@@ -163,6 +168,29 @@ def makeRequest_pycurl(self, uri=None, params={}, verb='GET',
+                # self.encode
+                encoded_data = self.encode(params)
+            if isinstance(encoded_data, basestring):
+                headers["Content-length"] = str(len(encoded_data))
```
Content-length need to be recalculated in param is dict. since param is passed below not encoded_data in line 194

ticoann commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -96,12 +98,20 @@ def set_opts(self, curl, url, params, headers,
             if params:
-                curl.setopt(pycurl.POSTFIELDS, json.dumps(params))
+                if isinstance(params, dict):
+                    curl.setopt(pycurl.POSTFIELDS, json.dumps(params))
```
content length need to be recalculated here.

ticoann commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -96,12 +98,20 @@ def set_opts(self, curl, url, params, headers,
             if params:
-                curl.setopt(pycurl.POSTFIELDS, json.dumps(params))
+                if isinstance(params, dict):
+                    curl.setopt(pycurl.POSTFIELDS, json.dumps(params))
```
content length need to be recalculated here.

ticoann commented:
Hi Valentin, I ran some test with this patch. Especially following unittest
https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py

It is successful without this patch but fails if I applied patch.
First problem is  the dbs validation fails. (am not sure why this fails here but not when httplib2 call is used) - it might be some authentication setting problem.
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/WMWorkloadTools.py#L67

Even if I ignore this error (by simply returning without calling DBS). It still failed put request with bad request error. It seems problem lies on encoding and content-type mismatch. I will try to debug more. But if you can look at that, I will appreciate. Thanks,

vkuznet commented:
Seangchan,
If time permit I'll have a look, but I doubt I'll be able to work until next
week. I saw your comments and I think you're right about Content-type.
Valentin.

On  0, ticoann <notifications@github.com> wrote:
> Hi Valentin, I ran some test with this patch. Especially following unittest
> https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py
> 
> It is successful without this patch but fails if I applied patch.
> First problem is  the dbs validation fails. (am not sure why this fails here but not when httplib2 call is used) - it might be some authentication setting problem.
> https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/WMWorkloadTools.py#L67
> 
> Even if I ignore this error (by simply returning without calling DBS). It still failed put request with bad request error. It seems problem lies on encoding and content-type mismatch. I will try to debug more. But if you can look at that, I will appreciate. Thanks, 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-167882960

ticoann commented:
test this please

ticoann commented:
Thanks Valentin, take your time. and happy new year.
Here is what I found. 
1. putTeam call is weird. It expects urlencoded data from url string but not body parts. 
   previous code append the params to url string if content_type is application/x-www-form-urlencoded. which I don't think correct way to do.
2.  There seems to be decoding twice if the pycurl is used.
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/RequestManager/RequestManager.py#L61
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/pycurl_manager.py#L143

previous versions default decode function just change to string not json.
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Requests.py#L294
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Requests.py#L280

vkuznet committed:
Add Content-length into pycurl POST/PUT requests

vkuznet commented:
Seangchan, I modified pycurl_manager to properly calculate Content-length. But during tests I found that RequestManager_t.py fails in setUp call since I don't have scramArch and CMSSWVersion parameters. I can't find where exactly they are setup, may be I'm lack of some config file in my test setup. Here is failed part: https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/RequestManager_t/utils.py#L25 and it reads schema from here https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/RequestManager_t/utils.py#L25 I can't find anywhere in WMCore where those parameters are set, so I modified utils.py to setup them manually and at least my RequestManager_t fails further down. The good sign that now it fails identically using pycurl and httplib.

vkuznet commented:
Hmm, seems like my last changes allow to pass all tests, am I right?

vkuznet committed:
Add zero content-length in POST request when no params is give; change logging report of failed json'ise to debug; remove obsolete code in Request for pycurl

vkuznet commented:
To make RequestManager_t unittest to pass we need https://github.com/dmwm/WMCore/pull/6495

ticoann commented:
test this please

ticoann commented:
@vkuznet, Hi Valentin, after applying #649, I am getting following errors when I run WMCore_t/Services_t/RequestManager_t/RequestManager_t.py test. Are you getting a same error?

```
DEBUG:cherrypy.error:PUT to request expects 4 argument(s), got 80
DEBUG:cherrypy.error:call to PUT with args: ['request'] kwargs: {} resulted in Invalid input: Input arguments failed sanitation.
INFO:cherrypy.access:[11/Jan/2016:09:59:37] mac-122010.dhcp.fnal.gov ::1 "PUT /unittests/rest/request HTTP/1.1" 400 Bad Request [data: - in 103 out 958 us ] [auth: - "" "" ] [ref: "" "WMCore.Services.Requests/v001" ]
Exception during set up, reason: Traceback (most recent call last):
  File "/Users/sryu/WorkArea2/WMCore/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py", line 66, in setUp
    r = self.jsonSender.put('request', schema)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Requests.py", line 130, in put
    encode, decode, contentType)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Requests.py", line 147, in makeRequest
    encoder, decoder, contentType)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Requests.py", line 173, in makeRequest_pycurl
    verb=verb, ckey=ckey, cert=cert, capath=capath, decode=decoder)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/pycurl_manager.py", line 201, in request
    raise exc
HTTPException: url=http://localhost:8899/unittests/rest/request, code=400, reason=Bad Request, headers={'Content-Length': '103', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Etag': '2211139151384643101', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Mon, 11 Jan 2016 15:59:37 GMT', 'Content-Type': 'application/json'}

E
======================================================================
ERROR: testA_RequestManagerService (__main__.RequestManagerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sryu/WorkArea2/WMCore/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py", line 74, in setUp
    raise ex
HTTPException: url=http://localhost:8899/unittests/rest/request, code=400, reason=Bad Request, headers={'Content-Length': '103', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Etag': '2211139151384643101', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Mon, 11 Jan 2016 15:59:37 GMT', 'Content-Type': 'application/json'}

----------------------------------------------------------------------
Ran 1 test in 1.236s

FAILED (errors=1)
```

vkuznet commented:
Seangchan,
my test is clean, I think I forgot to apply this part to RequestManager_t.py

 63         try:
 64             r = self.jsonSender.put('request/' + schema['RequestName'], schema)
 65             try:
 66                 self.requestName = r[0]['RequestName']
 67             except:
 68                 self.requestName = r[0].values()[0]['RequestName']
 69         except Exception as ex:
 70             print "Exception during set up, reason: %s" % ex
 71             raise ex

so, try to modify your line 66 with 

r = self.jsonSender.put('request/' + schema['RequestName'], schema)

the PUT request needs the name of request in URI !

Valentin.

On  0, ticoann <notifications@github.com> wrote:
> @vkuznet, Hi Valentin, after applying #649, I am getting following errors when I run WMCore_t/Services_t/RequestManager_t/RequestManager_t.py test. Are you getting a same error?
> 
> ```
> DEBUG:cherrypy.error:PUT to request expects 4 argument(s), got 80
> DEBUG:cherrypy.error:call to PUT with args: ['request'] kwargs: {} resulted in Invalid input: Input arguments failed sanitation.
> INFO:cherrypy.access:[11/Jan/2016:09:59:37] mac-122010.dhcp.fnal.gov ::1 "PUT /unittests/rest/request HTTP/1.1" 400 Bad Request [data: - in 103 out 958 us ] [auth: - "" "" ] [ref: "" "WMCore.Services.Requests/v001" ]
> Exception during set up, reason: Traceback (most recent call last):
>   File "/Users/sryu/WorkArea2/WMCore/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py", line 66, in setUp
>     r = self.jsonSender.put('request', schema)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Requests.py", line 130, in put
>     encode, decode, contentType)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Requests.py", line 147, in makeRequest
>     encoder, decoder, contentType)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Requests.py", line 173, in makeRequest_pycurl
>     verb=verb, ckey=ckey, cert=cert, capath=capath, decode=decoder)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/pycurl_manager.py", line 201, in request
>     raise exc
> HTTPException: url=http://localhost:8899/unittests/rest/request, code=400, reason=Bad Request, headers={'Content-Length': '103', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Etag': '2211139151384643101', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Mon, 11 Jan 2016 15:59:37 GMT', 'Content-Type': 'application/json'}
> 
> E
> ======================================================================
> ERROR: testA_RequestManagerService (__main__.RequestManagerTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/sryu/WorkArea2/WMCore/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py", line 74, in setUp
>     raise ex
> HTTPException: url=http://localhost:8899/unittests/rest/request, code=400, reason=Bad Request, headers={'Content-Length': '103', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Etag': '2211139151384643101', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Mon, 11 Jan 2016 15:59:37 GMT', 'Content-Type': 'application/json'}
> 
> ----------------------------------------------------------------------
> Ran 1 test in 1.236s
> 
> FAILED (errors=1)
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-170597531

ticoann commented:
Hi Valentin,
I am still getting errors like following.
Server side.
```
DEBUG:cherrypy.error:call to PUT with args: ['message', 'Taizong_160111_151802_3315'] kwargs: {} resulted in cannot parse JSON description: error 

                                              stack trace: Traceback (most recent call last):
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/WebTools/RESTApi.py", line 103, in _buildResponse
    args, kwargs)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/WebTools/RESTModel.py", line 117, in handler
    data = methodCall(*params, **kwargs)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/WebTools/RESTModel.py", line 200, in wrapper
    return func(**input_data)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/HTTPFrontEnd/RequestManager/ReqMgrRESTModel.py", line 618, in putMessage
    obj = JsonWrapper.loads( body )
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Wrappers/JsonWrapper/__init__.py", line 40, in loads
    raise ex
DecodeError: cannot parse JSON description: error
```
Client code
```
======================================================================
ERROR: testA_RequestManagerService (__main__.RequestManagerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sryu/WorkArea2/WMCore/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py", line 127, in testA_RequestManagerService
    self.reqService.sendMessage(requestName,"error")
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/RequestManager/RequestManager.py", line 157, in sendMessage
    contentType = 'application/json')
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/RequestManager/RequestManager.py", line 56, in _getResult
    verb = verb, contentType = contentType)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Service.py", line 212, in refreshCache
    self.getData(cachefile, url, inputdata, incoming_headers, encoder, decoder, verb, contentType)
  File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Service.py", line 319, in getData
    raise he
HTTPException: url=http://localhost:8899/unittests/rest/message/Taizong_160111_151802_3315, code=500, reason=Internal Server Error, headers={'Content-Length': '68', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Etag': '-1680540008677053678', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Mon, 11 Jan 2016 21:18:02 GMT', 'Content-Type': 'application/json'}

----------------------------------------------------------------------
Ran 1 test in 1.707s

FAILED (errors=1)
```

vkuznet committed:
Fix message request, it should be send as JSON rather than plan string

vkuznet committed:
Remove JSON dumps since it is not required anymore

vkuznet commented:
Seangchan,
my fault, this is because I forgot to update sendMessage API of RequestManager.
The original code sent wrong PUT request, i.e. it uses PUT with
application/json content-type while sends string message. I fixed it with
proper wrapper of messages in args. I updated my PR, see
https://github.com/vkuznet/WMCore/blob/pycurl-fix8/src/python/WMCore/Services/RequestManager/RequestManager.py#L145

With this fix the error you saw should disappear.
Valentin.

On  0, ticoann <notifications@github.com> wrote:
> Hi Valentin,
> I am still getting errors like following.
> Server side.
> ```
> DEBUG:cherrypy.error:call to PUT with args: ['message', 'Taizong_160111_151802_3315'] kwargs: {} resulted in cannot parse JSON description: error 
> 
>                                               stack trace: Traceback (most recent call last):
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/WebTools/RESTApi.py", line 103, in _buildResponse
>     args, kwargs)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/WebTools/RESTModel.py", line 117, in handler
>     data = methodCall(*params, **kwargs)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/WebTools/RESTModel.py", line 200, in wrapper
>     return func(**input_data)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/HTTPFrontEnd/RequestManager/ReqMgrRESTModel.py", line 618, in putMessage
>     obj = JsonWrapper.loads( body )
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Wrappers/JsonWrapper/__init__.py", line 40, in loads
>     raise ex
> DecodeError: cannot parse JSON description: error
> ```
> Client code
> ```
> ======================================================================
> ERROR: testA_RequestManagerService (__main__.RequestManagerTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/sryu/WorkArea2/WMCore/test/python/WMCore_t/Services_t/RequestManager_t/RequestManager_t.py", line 127, in testA_RequestManagerService
>     self.reqService.sendMessage(requestName,"error")
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/RequestManager/RequestManager.py", line 157, in sendMessage
>     contentType = 'application/json')
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/RequestManager/RequestManager.py", line 56, in _getResult
>     verb = verb, contentType = contentType)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Service.py", line 212, in refreshCache
>     self.getData(cachefile, url, inputdata, incoming_headers, encoder, decoder, verb, contentType)
>   File "/Users/sryu/WorkArea2/WMCore/src/python/WMCore/Services/Service.py", line 319, in getData
>     raise he
> HTTPException: url=http://localhost:8899/unittests/rest/message/Taizong_160111_151802_3315, code=500, reason=Internal Server Error, headers={'Content-Length': '68', 'Expires': 'Sun, 28 Jan 2007 00:00:00 GMT', 'Vary': 'Accept-Encoding', 'Server': 'CherryPy/3.2.2', 'Etag': '-1680540008677053678', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache, must-revalidate', 'Date': 'Mon, 11 Jan 2016 21:18:02 GMT', 'Content-Type': 'application/json'}
> 
> ----------------------------------------------------------------------
> Ran 1 test in 1.707s
> 
> FAILED (errors=1)
> ```
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-170693876

amaltaro commented:
Valentin, Seangchan, could you please summarize what the status of this PR is?
If this PR is still valid and we want to keep testing it, we need Valentin to fix the conflicts.

vkuznet commented:
Alan,
my understanding that pycurl patch conflicts with old ReqMgr code and the idea
was to post-pone it until we retire ReqMgr (since its code needs re-factoring
to use new pycurl). Let's wait what Seangchan will say.
Meanwhile, can you elaborate which conflicts are you talking about.
Thanks,
Valentin.

On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> Valentin, Seangchan, could you please summarize what the status of this PR is?
> If this PR is still valid and we want to keep testing it, we need Valentin to fix the conflicts.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6360#issuecomment-180335057

amaltaro commented:
Thanks, Valentin! I'm refering to merge conflicts, but if ReqMgr needs to go away first, then we better not spend time on it right now.

amaltaro commented:
Most - if not all - of this implementation has been merged through #9365
